### PR TITLE
SPARKC-95: Java API cleanup

### PR DESCRIPTION
**Pull request by jacek-lewandowski**

### CODE REVIEW

File: project/CassandraSparkBuild.scala [hunk 1]
```diff
@@ -184,7 +198,8 @@ object Dependencies {
   val metrics = Seq(Metrics.metricsCore, Metrics.metricsJson)
 
   val testKit = Seq(Test.akkaTestKit, Test.commonsIO, Test.junit,
```
Comment (RussellSpitzer) [inline project/CassandraSparkBuild.scala]: I think we are going to want to change the style on lists like this, @helena was suggestting 
 
    val testKit = Seq(
        elements, elements, elements
        more elements,
       )

And I think that looks pretty good as well

File: spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/RDDJavaFunctions.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+     * Any RDD which can be used to saveToCassandra can be used to joinWithCassandra as well as any RDD
+     * which only specifies the partition Key of a Cassandra Table. This method executes single partition
+     * requests against the Cassandra Table and accepts the functional modifiers that a normal
+     * {@link CassandraTableScanRDD} takes.
+     * <p/>
+     * By default this method only uses the Partition Key for joining but any combination of columns
+     * which are acceptable to C* can be used in the join. Specify columns using joinColumns as a
+     * parameter or the {@code on()} method.
+     */
+    public <R> CassandraJavaPairRDD<T, R> joinWithCassandraTable(
```
Comment (RussellSpitzer) [inline spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/RDDJavaFunctions.java]: Style, One param per line here
I think with invocation we break it up as well See Spark-99

File: spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/RDDJavaFunctions.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        ClassTag<T> classTagT = rdd.toJavaRDD().classTag();
+        ClassTag<R> classTagR = JavaApiHelper.getClassTag(rowReaderFactory.targetClass());
+
+        CassandraConnector connector = defaultConnector();
+        Option<ClusteringOrder> clusteringOrder = Option.empty();
+        Option<Object> limit = Option.empty();
+        CqlWhereClause whereClause = CqlWhereClause.empty();
+        ReadConf readConf = ReadConf.fromSparkConf(rdd.conf());
+
+        CassandraJoinRDD<T, R> joinRDD = new CassandraJoinRDD<>(rdd, keyspaceName, tableName,
```
Comment (RussellSpitzer) [inline spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/RDDJavaFunctions.java]: One param per line SPARKC-99

File: spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/rdd/CassandraJavaPairRDD.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+import com.datastax.spark.connector.cql.CassandraConnector;
+import com.datastax.spark.connector.japi.PairRDDJavaFunctions;
+import com.datastax.spark.connector.rdd.CassandraRDD;
+import com.datastax.spark.connector.rdd.ReadConf;
+import com.datastax.spark.connector.util.JavaApiHelper;
+
+@SuppressWarnings("UnusedDeclaration")
+public class CassandraJavaPairRDD<K, V> extends JavaPairRDD<K, V> {
+
+    public CassandraJavaPairRDD(CassandraRDD<Tuple2<K, V>> rdd,
```
Comment (RussellSpitzer) [inline spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/rdd/CassandraJavaPairRDD.java]: Style - One param per line

File: spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/rdd/CassandraJoinJavaRDD.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import scala.reflect.ClassTag;
+
+import static com.datastax.spark.connector.japi.CassandraJavaUtil.safeClassTag;
+
+import com.datastax.spark.connector.ColumnSelector;
+import com.datastax.spark.connector.rdd.CassandraJoinRDD;
+
+public class CassandraJoinJavaRDD<K, V> extends CassandraJavaPairRDD<K, V> {
+
+    public CassandraJoinJavaRDD(CassandraJoinRDD<K, V> rdd,
```
Comment (RussellSpitzer) [inline spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/rdd/CassandraJoinJavaRDD.java]: Style, One param per line

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraRDD.scala [hunk 6]
```diff
@@ -35,10 +35,10 @@ abstract class CassandraRDD[R](_sc: SparkContext, dep: Seq[Dependency[_]])(impli
 
   protected def connector: CassandraConnector
 
-  def toEmptyCassandraRDD(): EmptyCassandraRDD[R]
+  def toEmptyCassandraRDD: EmptyCassandraRDD[R]
 
   /** Allows to set custom read configuration, e.g. consistency level or fetch size. */
-  def withReadConf(readConf: ReadConf) = {
+  def withReadConf(readConf: ReadConf): this.type = {
```
Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraRDD.scala]: In SPARKC-99 this gets changed so this section will need a rebase

File: spark-cassandra-connector-demos/simple-demos/src/main/java/com/datastax/spark/connector/demo/JavaApiDemo.java [hunk 7]
```diff
@@ -85,7 +83,7 @@ public String call(Person person) throws Exception {
 
         // use case: we want to explicitly set a projection on the column set
         JavaRDD<String> rdd4 = javaFunctions(sc).cassandraTable("test", "people")
-                .select("id").map(new Function<CassandraRow, String>() {
+                .select(new String[] {"id"}).map(new Function<CassandraRow, String>() {
```
Comment (RussellSpitzer) [inline spark-cassandra-connector-demos/simple-demos/src/main/java/com/datastax/spark/connector/demo/JavaApiDemo.java]: With the varags back i'm not sure why this usage of select needs this syntax

File: spark-cassandra-connector-demos/simple-demos/src/main/java/com/datastax/spark/connector/demo/JavaApiDemo.java [hunk 8]
```diff
@@ -75,7 +73,7 @@ public String call(Person person) throws Exception {
 
         // use case: we want to filter rows on the database side with use of the where clause
         JavaRDD<String> rdd3 = javaFunctions(sc).cassandraTable("test", "people", mapRowTo(Person.class))
-                .where("name=?", "Anna").map(new Function<Person, String>() {
+                .where("name=?", new Object[] {"Anna"}).map(new Function<Person, String>() {
```
Comment (RussellSpitzer) [inline spark-cassandra-connector-demos/simple-demos/src/main/java/com/datastax/spark/connector/demo/JavaApiDemo.java]: Again not sure why this needs the Object[] when the var-arg method is below

File: project/CassandraSparkBuild.scala [hunk 9]
```diff
@@ -41,20 +48,25 @@ object CassandraSparkBuild extends Build {
...
 
   lazy val demos = RootProject("demos", demosPath, Seq(simpleDemos, kafkaStreaming, twitterStreaming))
 
-  lazy val simpleDemos = Project(
-    id = "simple-demos",
-    base = demosPath / "simple-demos",
-    settings = demoSettings,
-    dependencies = Seq(connector, jconnector, embedded)
+  lazy val simpleDemos =
+    Project(
```
Comment (RussellSpitzer) [inline project/CassandraSparkBuild.scala]: I'm not seeing the java api demo in this list anymore even when not using scala211

Comment (RussellSpitzer) [inline project/CassandraSparkBuild.scala]: 16:57:05 ➜  ~/repos/spark-cassandra-connector git:(SPARKC-95) ✗ sbt simple-demos/run
   [info] Loading project definition from /Users/russellspitzer/repos/spark-cassandra-connector/project
    
     Scala: 2.10.4 [To build against Scala 2.11 use '-Dscala-2.11=true']
     Scala Binary: 2.10
     Java: target=1.7 user=1.7.0_60

    [info] Set current project to root (in build file:/Users/russellspitzer/repos/spark-cassandra-connector/)
    [info] Compiling 1 Java source to /Users/russellspitzer/repos/spark-cassandra-connector/spark-cassandra-     connector-demos/simple-demos/target/scala-2.10/classes...
    [warn] Multiple main classes detected.  Run 'show discoveredMainClasses' to see the list

     Multiple main classes detected, select one to run:
     [1] com.datastax.spark.connector.demo.WordCountDemo
     [2] com.datastax.spark.connector.demo.TableCopyDemo
     [3] com.datastax.spark.connector.demo.AkkaStreamingDemo
     [4] com.datastax.spark.connector.demo.SQLDemo
     [5] com.datastax.spark.connector.demo.BasicReadWriteDemo

File: project/CassandraSparkBuild.scala [hunk 10]
```diff
@@ -41,20 +48,25 @@ object CassandraSparkBuild extends Build {
...
+      id = "simple-demos",
+      base = demosPath / "simple-demos",
+      settings = demoSettings ++ Seq(
+        excludeFilter in unmanagedSources := (if (scala211)
+          HiddenFileFilter || "*.java" else HiddenFileFilter)),
```
Comment (RussellSpitzer) [inline project/CassandraSparkBuild.scala]: I guess i'm not sure what this line is doing

    If (scala211) HiddenFileFilter || "*.java" else HiddenFileFilter

**Comment by RussellSpitzer** [comment]
Thanks for porting in the remaining Scala apis into Java!

A few things need to be tweaked, 
JavaApiDemo Code/Visitability

A few things should be looked into but are not critical
Style - Params/ Indentation and New lines

But I'm really happy that we can keep the standard api almost identical for end users

File: spark-cassandra-connector-demos/simple-demos/src/main/java/com/datastax/spark/connector/demo/JavaApiDemo.java (hunk 8 above)
Comment (jacek-lewandowski) [inline spark-cassandra-connector-demos/simple-demos/src/main/java/com/datastax/spark/connector/demo/JavaApiDemo.java]: it isn't - i forgot about this demo app

**Comment by jacek-lewandowski** [comment]
@RussellSpitzer i applied changes. 

For the problem that you do not see the demo app when compiling with Scala 2.10 - I don't know - could you retest? I tried `show discoveredMainClasses` and it this demo app was listed there.

File: doc/7_java_api.md [hunk 11]
```diff
@@ -284,4 +284,8 @@ Cassandra all the *RDDs* in that *DStream*.
 A longer example (with source code) of the Connector Java API is on the DataStax tech blog:
 [Accessing Cassandra from Spark in Java](http://www.datastax.com/dev/blog/accessing-cassandra-from-spark-in-java).
 
+### Scala 2.11
+
+Java API 1.2.0 is not compiled against Scala 2.11 because of https://issues.apache.org/jira/browse/SPARK-3266
```
Comment (pkolaczk) [inline doc/7_java_api.md]: [SPARK-3266](https://issues.apache.org/jira/browse/SPARK-3266)

File: project/CassandraSparkBuild.scala [hunk 12]
```diff
@@ -41,20 +50,25 @@ object CassandraSparkBuild extends Build {
...
-    id = s"$namespace-java",
-    base = file(s"$namespace-java"),
-    settings = connector.settings,
-    dependencies = Seq(connector % "compile;runtime->runtime;test->test;it->it,test;provided->provided")
+  lazy val jconnector = CrossScalaVersionsProject(
+    name = s"$namespace-java",
+    conf = assembledSettings ++ Seq(libraryDependencies ++= Dependencies.connector ++ Seq(
+        "org.scala-lang" % "scala-reflect"  % scalaVersion.value,
+        "org.scala-lang" % "scala-compiler" % scalaVersion.value % "test,it"), compileOrder in Compile := CompileOrder.ScalaThenJava)
+    ).copy(dependencies = Seq(connector % "compile;runtime->runtime;test->test;it->it,test;provided->provided")
```
Comment (pkolaczk) [inline project/CassandraSparkBuild.scala]: lines too long

Comment (pkolaczk) [inline project/CassandraSparkBuild.scala]: that's not just line too long, but this whole expression is hard to read; some parts of it should be taken out to another variable

File: spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/RDDJavaFunctions.java [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+    public CassandraConnector defaultConnector() {
+        return CassandraConnector$.MODULE$.apply(rdd.conf());
+    }
+
+    public SparkConf getConf() {
+        return rdd.conf();
+    }
+
+    public void saveToCassandra(String keyspace,
```
Comment (pkolaczk) [inline spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/RDDJavaFunctions.java]: Indentation:
```scala
    public void saveToCassandra(
            String keyspace,
            String table,
            RowWriterFactory<T> rowWriterFactory,
            ColumnSelector columnNames,
            WriteConf conf,
            CassandraConnector connector) {
```

File: spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/RDDJavaFunctions.java (hunk 2 above)
Comment (pkolaczk) [inline spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/RDDJavaFunctions.java]: Yes, with long invocations, we should put one param per line.

File: spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/RDDJavaFunctions.java [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+                selectedColumns,
+                joinColumns,
+                whereClause,
+                limit,
+                clusteringOrder,
+                readConf,
+                classTagT,
+                classTagR,
+                rowWriterFactory,
+                rowReaderFactory);
```
Comment (pkolaczk) [inline spark-cassandra-connector-java/src/main/java/com/datastax/spark/connector/japi/RDDJavaFunctions.java]: This is something I'd like to fix in the next round of code cleanup. Those parameter lists are getting crazy.

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/JavaApiHelper.scala [hunk 15]
```diff
@@ -42,6 +45,19 @@ object JavaApiHelper {
...
+  def valuesAsJavaIterable[K, V, IV <: Iterable[V]]: ((K, IV)) => (K, JIterable[V]) = {
+    case (k, iterable) => (k, asJavaIterable(iterable))
+  }
+
+  def valuesAsJavaCollection[K, V, IV <: Iterable[V]]: ((K, IV)) => (K, JCollection[V]) = {
+    case (k, iterable) => (k, asJavaCollection(iterable))
+  }
+
   /** Returns a runtime class of a given `TypeTag`. */
   def getRuntimeClass[T](typeTag: TypeTag[T]): Class[T] = mirror.runtimeClass(typeTag.tpe).asInstanceOf[Class[T]]
```
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/JavaApiHelper.scala]: line too long, can be easily split after =

**Comment by pkolaczk** [comment]
Overall looks much better than it was before
+1

**Comment by RussellSpitzer** [comment]
Retested 

    commit 2776752485c7ca345e4a342aa3e061b3b7b1b6f9
    Author: Jacek Lewandowski <lewandowski.jacek@gmail.com>
    Date:   Wed Mar 25 15:40:24 2015 +0100

    SPARKC-95: Addressed reviewer comments


8:39:30 ➜  ~/repos/spark-cassandra-connector git:(SPARKC-95) ✗ ./sbt/sbt simple-demos/run
Launching sbt from sbt/sbt-launch-0.13.8.jar
[info] Loading project definition from /Users/russellspitzer/repos/spark-cassandra-connector/project

  Scala: 2.10.4 [To build against Scala 2.11 use '-Dscala-2.11=true']
  Scala Binary: 2.10
  Java: target=1.7 user=1.7.0_60

[info] Set current project to root (in build file:/Users/russellspitzer/repos/spark-cassandra-connector/)
[warn] Multiple main classes detected.  Run 'show discoveredMainClasses' to see the list

Multiple main classes detected, select one to run:

 [1] com.datastax.spark.connector.demo.AkkaStreamingDemo
 [2] com.datastax.spark.connector.demo.BasicReadWriteDemo
 [3] com.datastax.spark.connector.demo.SQLDemo
 [4] com.datastax.spark.connector.demo.TableCopyDemo
 [5] com.datastax.spark.connector.demo.WordCountDemo

Enter number:

**Comment by jacek-lewandowski** [comment]
@RussellSpitzer I don't really know how it is possible:

```
[info] Loading global plugins from /Users/jlewandowski/.sbt/0.13/plugins
[info] Loading project definition from /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/project

  Scala: 2.10.4 [To build against Scala 2.11 use '-Dscala-2.11=true']
  Scala Binary: 2.10
  Java: target=1.7 user=1.7.0_60
        
[info] Set current project to root (in build file:/Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/)
 SBT   SPARKC-95  root  clean
[success] Total time: 0 s, completed Mar 25, 2015 6:15:19 PM
 SBT   SPARKC-95  root  package
[info] Updating {file:/Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/}root...
[info] Done updating.
[info] Updating {file:/Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/}demos...
[info] Done updating.
[info] Compiling 11 Scala sources to /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/spark-cassandra-connector-embedded/target/scala-2.10/classes...
[info] Compiling 128 Scala sources and 1 Java source to /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/spark-cassandra-connector/target/scala-2.10/classes...
[warn] Class org.apache.cassandra.io.util.DataOutputPlus not found - continuing with a stub.
[warn] Class org.apache.cassandra.io.util.DataOutputPlus not found - continuing with a stub.
[warn] non-nullary method overrides nullary method
[warn] three warnings found
[info] Packaging /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/spark-cassandra-connector-embedded/target/scala-2.10/spark-cassandra-connector-embedded_2.10-1.2.0-SNAPSHOT.jar ...
[info] Done packaging.
[info] Packaging /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/target/scala-2.10/root_2.10-1.2.0-SNAPSHOT.jar ...
[info] Done packaging.
[info] Packaging /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/spark-cassandra-connector-demos/target/scala-2.10/demos_2.10-1.2.0-SNAPSHOT.jar ...
[info] Done packaging.
[info] Packaging /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/spark-cassandra-connector/target/scala-2.10/spark-cassandra-connector_2.10-1.2.0-SNAPSHOT.jar ...
[info] Done packaging.
[info] Compiling 4 Scala sources and 12 Java sources to /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/spark-cassandra-connector-java/target/scala-2.10/classes...
[info] Compiling 1 Scala source to /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/spark-cassandra-connector-demos/kafka-streaming/target/scala-2.10/classes...
[info] Compiling 3 Scala sources to /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/spark-cassandra-connector-demos/twitter-streaming/target/scala-2.10/classes...
[info] Packaging /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/spark-cassandra-connector-java/target/scala-2.10/spark-cassandra-connector-java_2.10-1.2.0-SNAPSHOT.jar ...
[info] Done packaging.
[info] Packaging /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/spark-cassandra-connector-demos/kafka-streaming/target/scala-2.10/kafka-streaming_2.10-1.2.0-SNAPSHOT.jar ...
[info] Done packaging.
[info] Packaging /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/spark-cassandra-connector-demos/twitter-streaming/target/scala-2.10/twitter-streaming_2.10-1.2.0-SNAPSHOT.jar ...
[info] Done packaging.
[info] Compiling 7 Scala sources and 1 Java source to /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/spark-cassandra-connector-demos/simple-demos/target/scala-2.10/classes...
[warn] Multiple main classes detected.  Run 'show discoveredMainClasses' to see the list
[info] Packaging /Users/jlewandowski/Projects/OpenSource/spark-cassandra-connector/spark-cassandra-connector-demos/simple-demos/target/scala-2.10/simple-demos_2.10-1.2.0-SNAPSHOT.jar ...
[info] Done packaging.
[success] Total time: 41 s, completed Mar 25, 2015 6:16:01 PM
 SBT   SPARKC-95  root  simple-demos/run
[warn] Multiple main classes detected.  Run 'show discoveredMainClasses' to see the list

Multiple main classes detected, select one to run:

 [1] com.datastax.spark.connector.demo.AkkaStreamingDemo
 [2] com.datastax.spark.connector.demo.BasicReadWriteDemo
 [3] com.datastax.spark.connector.demo.JavaApiDemo
 [4] com.datastax.spark.connector.demo.SQLDemo
 [5] com.datastax.spark.connector.demo.TableCopyDemo
 [6] com.datastax.spark.connector.demo.WordCountDemo

Enter number: 
```

**Pull request merged**